peps/pep-0667.rst

```diff
@@ -40,72 +62,227 @@ The current implementation of ``locals()`` and ``frame.f_locals`` is slow,
 inconsistent and buggy.
 We want to make it faster, consistent, and most importantly fix the bugs.
 
-For example::
+For example, when attempting to manipulate local variables via frame objects::
 
     class C:
         x = 1
         sys._getframe().f_locals['x'] = 2
         print(x)
 
-prints ``2``
-
-but::
+prints ``2``, but::
 
     def f():
         x = 1
         sys._getframe().f_locals['x'] = 2
         print(x)
     f()
 
-prints ``1``
+prints ``1``.
 
-This is inconsistent, and confusing.
-With this PEP both examples would print ``2``.
+This is inconsistent, and confusing. Worse than that, the current behavior can
+result in strange `bugs <https://github.com/python/cpython/issues/74929>`__.
 
-Worse than that, the current behavior can result in strange `bugs
-<https://github.com/python/cpython/issues/74929>`__.
+With this PEP both examples would print ``2`` as the function level
+change would be written directly to the optimized local variables in
+the frame rather than to a cached dictionary snapshot.
 
 There are no compensating advantages for the current behavior;
 it is unreliable and slow.
 
+The ``locals()`` builtin has its own undesirable behaviours, where
+adding a new local variable can change the behaviour of code
+executed with ``exec()`` in function scopes, even if that code
+runs *before* the local variable is defined.
+
+For example::
+
+    def f():
+        exec("x = 1")
+        print(locals().get("x"))
+    f()
+
+prints ``1``, but::
+
+    def f():
+        exec("x = 1")
+        print(locals().get("x"))
+        x = 0
+    f()
+
+prints ``None`` (the default value from the ``.get()`` call).
+
+With this PEP both examples would print ``None``, as the call to
+``exec()`` and the subsequent call to ``locals()`` would use
+independent dictionary snapshots of the local variables rather
+than using the same shared dictionary cached on the frame object.
+
 Rationale
 =========
 
+Making the ``frame.f_locals`` attribute a write-through proxy
+-------------------------------------------------------------
+
 The current implementation of ``frame.f_locals`` returns a dictionary
```

Comment:
"current" could be ambiguous in the future when read with historical context.  Five years from now, what does "current" mean?  Could you be more explicit?